File: win-prog/ProgWinCS/Printing/PrintThreePages/PrintThreePages.cs
//----------------------------------------------
// PrintThreePages.cs © 2001 by Charles Petzold
//----------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Printing;
using System.Windows.Forms;

class PrintThreePages: Form
{
     const int iNumberPages = 3;
     int       iPageNumber;

     public static void Main()
     {
          Application.Run(new PrintThreePages());
     }
     public PrintThreePages()
     {
          Text = "Print Three Pages";

          Menu = new MainMenu();
          Menu.MenuItems.Add("&File");
          Menu.MenuItems[0].MenuItems.Add("&Print...", 
                                   new EventHandler(MenuFilePrintOnClick));
     }
     void MenuFilePrintOnClick(object obj, EventArgs ea)
     {
               // Create PrintDocument.

          PrintDocument prndoc = new PrintDocument();

               // Create dialog box and set PrinterName property.

          PrinterSelectionDialog dlg = new PrinterSelectionDialog();
          dlg.PrinterName = prndoc.PrinterSettings.PrinterName;

               // Show dialog box and bail out if not OK.

          if (dlg.ShowDialog() != DialogResult.OK)
               return;

               // Set PrintDocument to selected printer.

          prndoc.PrinterSettings.PrinterName = dlg.PrinterName;

               // Set printer resolution to "draft".

          foreach (PrinterResolution prnres in 
                                   prndoc.PrinterSettings.PrinterResolutions)
          {
               if (prnres.Kind == PrinterResolutionKind.Draft)
                    prndoc.DefaultPageSettings.PrinterResolution = prnres;
          }

               // Set remainder of PrintDocument properties.

          prndoc.DocumentName = Text;
          prndoc.PrintPage += new PrintPageEventHandler(OnPrintPage);
          prndoc.QueryPageSettings += new QueryPageSettingsEventHandler
                                             (OnQueryPageSettings);
               // Commence printing.

          iPageNumber = 1;
          prndoc.Print();
     }
     void OnQueryPageSettings(object obj, QueryPageSettingsEventArgs qpsea)
     {
          if (qpsea.PageSettings.PrinterSettings.LandscapeAngle != 0)
               qpsea.PageSettings.Landscape ^= true;
     }
     void OnPrintPage(object obj, PrintPageEventArgs ppea)
     {
          Graphics grfx  = ppea.Graphics;
          Font     font  = new Font("Times New Roman", 360);
          string   str   = iPageNumber.ToString();
          SizeF    sizef = grfx.MeasureString(str, font);

          grfx.DrawString(str, font, Brushes.Black, 
                    (grfx.VisibleClipBounds.Width - sizef.Width) / 2,
                    (grfx.VisibleClipBounds.Height - sizef.Height) / 2);

          ppea.HasMorePages = iPageNumber < iNumberPages;
          iPageNumber += 1;
     }
}

File: win-prog/ProgWinVB/Printing/PrintThreePages/PrintThreePages.vb
'------------------------------------------------
' PrintThreePages.vb (c) 2002 by Charles Petzold
'------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Printing
Imports System.Windows.Forms

Class PrintThreePages
    Inherits Form

    Const iNumberPages As Integer = 3
    Private iPageNumber As Integer

    Shared Sub Main()
        Application.Run(New PrintThreePages())
    End Sub

    Sub New()
        Text = "Print Three Pages"

        Menu = New MainMenu()
        Menu.MenuItems.Add("&File")
        Menu.MenuItems(0).MenuItems.Add("&Print...", _
                                        AddressOf MenuFilePrintOnClick)
    End Sub

    Private Sub MenuFilePrintOnClick(ByVal obj As Object, _
                                     ByVal ea As EventArgs)
        ' Create PrintDocument.
        Dim prndoc As New PrintDocument()

        ' Create dialog box and set PrinterName property.
        Dim dlg As New PrinterSelectionDialog()
        dlg.PrinterName = prndoc.PrinterSettings.PrinterName

        ' Show dialog box and bail out if not OK.
        If dlg.ShowDialog() <> DialogResult.OK Then Return

        ' Set PrintDocument to selected printer.
        prndoc.PrinterSettings.PrinterName = dlg.PrinterName

        ' Set printer resolution to "draft".
        Dim prnres As PrinterResolution
        For Each prnres In prndoc.PrinterSettings.PrinterResolutions
            If prnres.Kind = PrinterResolutionKind.Draft Then
                prndoc.DefaultPageSettings.PrinterResolution = prnres
                Exit For
            End If
        Next

        ' Set remainder of PrintDocument properties.
        prndoc.DocumentName = Text
        AddHandler prndoc.PrintPage, AddressOf OnPrintPage
        AddHandler prndoc.QueryPageSettings, AddressOf OnQueryPageSettings

        ' Commence printing.
        iPageNumber = 1
        prndoc.Print()
    End Sub

    Private Sub OnQueryPageSettings(ByVal obj As Object, _
                                ByVal qpsea As QueryPageSettingsEventArgs)
        If qpsea.PageSettings.PrinterSettings.LandscapeAngle <> 0 Then
            qpsea.PageSettings.Landscape = Not qpsea.PageSettings.Landscape
        End If
    End Sub

    Private Sub OnPrintPage(ByVal obj As Object, _
                            ByVal ppea As PrintPageEventArgs)
        Dim grfx As Graphics = ppea.Graphics
        Dim fnt As New Font("Times New Roman", 360)
        Dim str As String = iPageNumber.ToString()
        Dim szf As SizeF = grfx.MeasureString(str, fnt)

        grfx.DrawString(str, fnt, Brushes.Black, _
                        (grfx.VisibleClipBounds.Width - szf.Width) / 2, _
                        (grfx.VisibleClipBounds.Height - szf.Height) / 2)
        ppea.HasMorePages = iPageNumber < iNumberPages
        iPageNumber += 1
    End Sub
End Class
